### DISCUSSION TITLE
Timeout unexpectedly not raised

### DISCUSSION THREAD

**Discussion by valsteen** [encode/httpcore#814] [Potential Issue] [answered]
Hi, this is a question I originally asked in httpx: https://github.com/encode/httpx/discussions/2861

If you don't mind, since the same conditions apply, I paste again the question below, on the bottom I will add my findings in httpcore:

------

I was investigating why I sometimes have requests that take much longer than the timeout. I could isolate and reproduce the scenario:

I have those versions, on Python 3.11.3:

```
httpx 0.25.0 The next generation HTTP client.
├── certifi *
├── httpcore >=0.18.0,<0.19.0
│   ├── anyio >=3.0,<5.0 
│   │   ├── idna >=2.8 
│   │   └── sniffio >=1.1 
│   ├── certifi * 
│   ├── h11 >=0.13,<0.15 
│   └── sniffio ==1.* (circular dependency aborted here)
├── idna *
└── sniffio *
```

let's have this minimal server which just sleeps 2 seconds

```python
import asyncio
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root():
    await asyncio.sleep(2)
```

I name it `simple.py` and run `uvicorn simple:app`

And here is the test:

```python
import asyncio
import time

from httpx import Limits, AsyncClient, Timeout


async def read():
    start = time.time()
    try:
        response = await http.get("http://127.0.0.1:8000")
        response.raise_for_status()
    except Exception as ex:
        print(f"failed: {ex.__class__}")

    print(f"time: {time.time()-start}")


async def main():
    tasks = []
    for x in range(100):
        tasks.append(asyncio.create_task(read()))
    await asyncio.gather(*tasks, return_exceptions=True)


if __name__ == "__main__":
    limits = Limits(
        max_keepalive_connections=None,
        max_connections=3,
    )
    http = AsyncClient(
        limits=limits,
        timeout=Timeout(5, read=5, pool=5),
    )

    asyncio.get_event_loop().run_until_complete(main())
```

I get timings way above 5 seconds, but I don't understand why ; I would expect a pool timeout.

I can probably work around this by having a timeout on the task itself ( like here: `asyncio.wait_for(read(), 5)` ), but did I misunderstand the functionality, or is this scenario not supported ?

------

Following this, I tracked down where the PoolTimeout logic originates from, and it's this `status.wait_for_connection(timeout=timeout)`

https://github.com/encode/httpcore/blob/94ffb33cf378e41662c90615458db9e42b43cad5/httpcore/_async/connection_pool.py#L229-L234

after that, it can fail with `ConnectionNotAvailable`

https://github.com/encode/httpcore/blob/94ffb33cf378e41662c90615458db9e42b43cad5/httpcore/_async/connection_pool.py#L246-L250

but then it loops and tries again with the same timeout, and depending on the size of the backlog this can go way beyond the intended timeout.

To validate this assumption, I tried this very rough patch:

```python
        started = time.time()

        while True:
            timeouts = request.extensions.get("timeout", {})
            timeout = timeouts.get("pool", None)

            timeout = max(0, timeout - (time.time() - started))

            try:
                connection = await status.wait_for_connection(timeout=timeout)
```

Following that, the example worked as I expected.

I presume this is not intentional, but please correct me if I'm wrong.

**Comment by zanieb** [comment]
Thanks for the thorough description! This looks like a bug to me. If another maintainer agrees we can work on a fix.

**Comment by karpetrosyan** [comment]
Hi @valsteen!

Thank you for providing such a clear description.
We actually expect this behavior, and it applies not only to the pool, but also to read/write operations (waiting longer than the timeout).

You can see #732 where we discussed what to do in such cases for read/write timeouts, but in the end we decided to keep it simple and not change anything ( Refs: [OverallTimeout class](https://github.com/encode/httpcore/pull/714/files#diff-d3a40cf09f980d710bfc8d32385d11086ebd179cec344a9ac5c095251f7b6224R21-R54) ).

Applying your solution appears reasonable to me, but we may encounter the same issues with read/write operations. For example, setting the read timeout to 5 may cause you to hang indefinitely because the server may send 1 byte every 4 seconds.

> I get timings way above 5 seconds, but I don't understand why ; I would expect a pool timeout.

We should probably resolve this if the difference is significant enough.

**Reply by valsteen** [reply]
Thank you,

for an outsider it's a bit hard to extract what exactly happened in #732 but I suppose I get the idea - the timeout is really more about fine-tuning the detection of stale connections than any guarantee of maximum total processing time.

here is a bit more context about the situation I try to solve:

We've got an API that calls a service that most of the time gives us <150ms response time, but there can be outliers taking several seconds.

In order to not reach a limit of maximum open files a shared httpx AsyncClient is initialized with max_connections, and to prevent abnormaly slow requests from blocking everyone else, there's a timeout of 15 seconds ( applies to all: read, write, pool ).

One day we get a pod crashing due to OOM. Below each spike happens right before a retart:

<img width="996" alt="image" src="https://github.com/encode/httpcore/assets/382143/71db9b1e-3369-4c36-aeeb-f4892a573c3f">


From there, I investigated further the application's behaviour and this was my surprise: despite the timeout, some requests take up to 8 minutes. If I count the amount of requests that exceeds 60 seconds, I get a huge spike of up to 15K http calls.   

![image](https://github.com/encode/httpcore/assets/382143/defabe3f-2764-472d-8077-dd9aacbffb92)

I thought at first there was a problem with our instrumentation or metrics, and this is were I tried the scenario above where I could exceed the given pool timeout. Given a sufficient amount of concurrent slow reads and a max_connection setting, new requests are blocked in a loop trying to acquire a connection from the pool, fail to do so, and try again waiting for the same time. So many requests get queued that it fills up memory and we get a OOM.

I'm aware that ultimately, controlling resource usage should be done at infrastructure level, something like 1/ the readiness probe should fail if the pool is full, so new requests don't get routed to that pod 2/ it should autoscale so there is always a connection readily available.

Still, if the documentation of an open source library gives me false impressions, I'd better report it so it can be improved for everyone else. It can be just a mention in the doc saying that the timeout is not a guarantee ; that cuts short debates among teams using it: just point out the doc, and proceed with appropriate measures - such as using a task with a deadline, knowing that it's perfectly expected and not due to a misuse.

Lastly it seems that considering the pool timeout similar to read/write timeout sounds debatable. Single byte read/write resets the timeout for that particular request, albeit still taking up a connection from the pool, but that's a limited problem if incoming requests fail early when there is no available connection. On the other hand, acquiring a connection from the pool is independent from the request you're willing to make, you just fail to acquire it because someone else took it. It begs the question: given enough concurrent requests, the pool timeout will become ineffective, so boldly said, why does it exist ?

**Reply by GalaxySnail** [reply]
> the timeout is really more about fine-tuning the detection of stale connections than any guarantee of maximum total processing time.

That's right. If you want something to limit the total processing time, [`trio.move_on_after()`](https://trio.readthedocs.io/en/latest/reference-core.html#cancellation-and-timeouts) or [`anyio.move_on_after()`](https://anyio.readthedocs.io/en/stable/cancellation.html) can be helpful. For details, please refer to: https://vorpus.org/blog/timeouts-and-cancellation-for-humans/

**Reply by karpetrosyan** [reply]
> I thought at first there was a problem with our instrumentation or metrics, and this is were I tried the scenario above where I could exceed the given pool timeout. Given a sufficient amount of concurrent slow reads and a max_connection setting, new requests are blocked in a loop trying to acquire a connection from the pool, fail to do so, and try again waiting for the same time. So many requests get queued that it fills up memory and we get a OOM.

> I'm aware that ultimately, controlling resource usage should be done at infrastructure level, something like 1/ the readiness probe should fail if the pool is full, so new requests don't get routed to that pod 2/ it should autoscale so there is always a connection readily available.

> after that, it can fail with ConnectionNotAvailable

It's very interesting; we do not expect the connection to wait for a stream for such a long time. We also do not expect the exception `ConnectionNotAvailable` to be frequent.
 
> From there, I investigated further the application's behaviour and this was my surprise: despite the timeout, some requests take up to 8 minutes. If I count the amount of requests that exceeds 60 seconds, I get a huge spike of up to 15K http calls.

For sure, if the connection could hang in the connection pool for 8 minutes, we should probably apply the solution that you have shown.

> Still, if the documentation of an open source library gives me false impressions, I'd better report it so it can be improved for everyone else. It can be just a mention in the doc saying that the timeout is not a guarantee ; that cuts short debates among teams using it: just point out the doc, and proceed with appropriate measures - such as using a task with a deadline, knowing that it's perfectly expected and not due to a misuse.

I completely agree with you. This should be mentioned in the documentation (if it's not going to be changed).

**Reply by valsteen** [reply]
@GalaxySnail yes thank you I'm aware of this. On my side I had this suggestion of custom transport using `wait_for` ( which for the sake of event loop compatibility, should probably use anyio instead of asyncio ) : https://github.com/encode/httpx/discussions/2861#discussioncomment-7120715

**Reply by karpetrosyan** [reply]
@valsteen 
Can we have a minimum code example to reproduce this issue?

**Reply by valsteen** [reply]
@karpetrosyan the first message of this discussion has a reproducible example, is there a problem with it ?

**Comment by karpetrosyan** [comment]
There is the same example that @valsteen has provided, but just using the `httpcore` library and using `trio` instead of `asyncio` like we do for our issues.

**file**: test.py
```python
import time
import trio
from httpcore import AsyncConnectionPool

TIMEOUTS = {"read": 5, "write": 5, "pool": 5, "connect": 5}


async def read():
    start = time.time()
    try:
        response = await http.request(
            "GET", "http://127.0.0.1:8000", extensions={"timeout": TIMEOUTS}
        )
        if response.status // 100 != 2:
            raise RuntimeError
    except Exception as ex:
        print(f"failed: {ex.__class__}")

    print(f"time: {time.time()-start}")


async def main():
    async with trio.open_nursery() as nursery:
        for _ in range(100):
            nursery.start_soon(read)


if __name__ == "__main__":
    http = AsyncConnectionPool(
        max_connections=3,
        max_keepalive_connections=3,
    )
    trio.run(main)
```

**Comment by valsteen** [comment, marked as answer]
this is solved:

- issue: https://github.com/encode/httpcore/issues/822
- PR : https://github.com/encode/httpcore/pull/823
